### Step 1

### .Importig Libraries

In [3]:
# Import libraries
import pandas as pd
import numpy as np
import os

### .Importing Data

In [5]:
# path
path = r"C:\Users\nabar\OneDrive\Documents\03-2025 Instacart Basket Analysis"
ords_prods_merge= pd.read_pickle(os.path.join(path, '02 Data' , 'Prepared Data' , 'ords_prods_merge_2.pkl'))

### Step 2

In [7]:
# Use groupby and agg functions to calculate mean of user_order_number for entire dataframe
ords_prods_merge.groupby('department_id').agg({'order_number': ['mean']})

order_number
                      mean
department_id             
1                15.457838
2                17.277920
3                17.170395
4                17.811403
5                15.215751
6                16.439806
7                17.225802
8                15.340650
9                15.895474
10               20.197148
11               16.170638
12               15.887671
13               16.583536
14               16.773669
15               16.165037
16               17.665606
17               15.694469
18               19.310397
19               17.177343
20               16.473447
21               22.902379

### Step 3

The full dataset contains 21 department IDs, while the subset of 100,000 records includes only 8 departments. Additionally, the mean order numbers vary across departments, showing differences between the subset and the complete dataset. This indicates that the subset may not fully represent the trends observed in the entire data frame.

### Step 4

In [11]:
# Creating new column "max_order"
ords_prods_merge['max_order'] = ords_prods_merge.groupby(['user_id'])['order_number'].transform("max")

In [12]:
ords_prods_merge.head(15)

order_id  user_id eval_set  order_number  orders_day_of_the_week  \
0    2539329        1    prior             1                       2   
1    2539329        1    prior             1                       2   
2    2539329        1    prior             1                       2   
3    2539329        1    prior             1                       2   
4    2539329        1    prior             1                       2   
5    2398795        1    prior             2                       3   
6    2398795        1    prior             2                       3   
7    2398795        1    prior             2                       3   
8    2398795        1    prior             2                       3   
9    2398795        1    prior             2                       3   
10   2398795        1    prior             2                       3   
11    473747        1    prior             3                       3   
12    473747        1    prior             3                       3   
13    473747        1    prior             3                       3   
14    473747        1    prior             3                       3   

    order_hour_of_day  days_since_prior_order  product_id  add_to_cart_order  \
0                   8                     NaN         196                  1   
1                   8                     NaN       14084                  2   
2                   8                     NaN       12427                  3   
3                   8                     NaN       26088                  4   
4                   8                     NaN       26405                  5   
5                   7                    15.0         196                  1   
6                   7                    15.0       10258                  2   
7                   7                    15.0       12427                  3   
8                   7                    15.0       13176                  4   
9                   7                    15.0       26088                  5   
10                  7                    15.0       13032                  6   
11                 12                    21.0         196                  1   
12                 12                    21.0       12427                  2   
13                 12                    21.0       10258                  3   
14                 12                    21.0       25133                  4   

    reordered                             product_name  aisle_id  \
0           0                                     Soda        77   
1           0  Organic Unsweetened Vanilla Almond Milk        91   
2           0                      Original Beef Jerky        23   
3           0               Aged White Cheddar Popcorn        23   
4           0         XL Pick-A-Size Paper Towel Rolls        54   
5           1                                     Soda        77   
6           0                               Pistachios       117   
7           1                      Original Beef Jerky        23   
8           0                   Bag of Organic Bananas        24   
9           1               Aged White Cheddar Popcorn        23   
10          0                    Cinnamon Toast Crunch       121   
11          1                                     Soda        77   
12          1                      Original Beef Jerky        23   
13          1                               Pistachios       117   
14          0                    Organic String Cheese        21   

    department_id  prices match    price_range_loc     busiest_day  \
0               7     9.0  both  Mid-range product  Regularly busy   
1              16    12.5  both  Mid-range product  Regularly busy   
2              19     4.4  both  Low-range product  Regularly busy   
3              19     4.7  both  Low-range product  Regularly busy   
4              17     1.0  both  Low-range product  Regularly busy   
5               7     9.0  both  Mid-range product  Regularly

### Creat Loyalty Flag

In [14]:
ords_prods_merge.loc[ords_prods_merge['max_order'] > 40, 'loyalty_flag'] = 'Loyal customer'

In [15]:
ords_prods_merge.loc[(ords_prods_merge['max_order'] <= 40) & (ords_prods_merge['max_order'] > 10), 'loyalty_flag'] = 'Regular customer'

In [16]:
ords_prods_merge.loc[ords_prods_merge['max_order'] <= 10, 'loyalty_flag'] = 'New customer'

In [17]:
# Check loyalty flags
ords_prods_merge['loyalty_flag'].value_counts(dropna = False)

loyalty_flag
Regular customer    15876776
Loyal customer      10284093
New customer         6243990
Name: count, dtype: int64

In [30]:
# Check Loyalty flag in dataframe
ords_prods_merge[['user_id', 'loyalty_flag', 'order_number','max_order']].head(30)

user_id  loyalty_flag  order_number  max_order
0         1  New customer             1         10
1         1  New customer             1         10
2         1  New customer             1         10
3         1  New customer             1         10
4         1  New customer             1         10
5         1  New customer             2         10
6         1  New customer             2         10
7         1  New customer             2         10
8         1  New customer             2         10
9         1  New customer             2         10
10        1  New customer             2         10
11        1  New customer             3         10
12        1  New customer             3         10
13        1  New customer             3         10
14        1  New customer             3         10
15        1  New customer             3         10
16        1  New customer             4         10
17        1  New customer             4         10
18        1  New customer             4         10
19        1  New customer             4         10
20        1  New customer             4         10
21        1  New customer             5         10
22        1  New customer             5         10
23        1  New customer             5         10
24        1  New customer             5         10
25        1  New customer             5         10
26        1  New customer             5         10
27        1  New customer             5         10
28        1  New customer             5         10
29        1  New customer             6         10

### Step 5

In [33]:
ords_prods_merge.groupby(['loyalty_flag']).agg({'prices':['mean','min','max']})

prices              
                       mean  min      max
loyalty_flag                             
Loyal customer    10.386336  1.0  99999.0
New customer      13.294670  1.0  99999.0
Regular customer  12.495717  1.0  99999.0

Loyal Customers: They tend to buy lower-cost items with an average price of $10.39.

New Customers: They spend the most, averaging $13.29, potentially exploring higher-priced products.

Regular Customers: They fall in the middle, with an average price of $12.50.

The minimum price of 1 doller is consistent across all groups, while the maximum price of $99,999 is likely an outlier or data error, skewing the analysis.

### Step 6

In [49]:
# Creating new column 'average_spend'
ords_prods_merge['average_spend']=ords_prods_merge.groupby(['user_id'])['prices'].transform("mean")

In [47]:
# Creating flag 
ords_prods_merge.loc[ords_prods_merge['average_spend']<10,'spender_flag']='Low_spender'

ords_prods_merge.loc[ords_prods_merge['average_spend']>= 10, 'spender_flag'] = 'High_spender'

In [51]:
# Check spender flags
ords_prods_merge['spender_flag'].value_counts(dropna = False)

spender_flag
Low_spender     31770614
High_spender      634245
Name: count, dtype: int64

In [53]:
ords_prods_merge[['user_id', 'average_spend', 'spender_flag']].head(20)

user_id  average_spend spender_flag
0         1       6.367797  Low_spender
1         1       6.367797  Low_spender
2         1       6.367797  Low_spender
3         1       6.367797  Low_spender
4         1       6.367797  Low_spender
5         1       6.367797  Low_spender
6         1       6.367797  Low_spender
7         1       6.367797  Low_spender
8         1       6.367797  Low_spender
9         1       6.367797  Low_spender
10        1       6.367797  Low_spender
11        1       6.367797  Low_spender
12        1       6.367797  Low_spender
13        1       6.367797  Low_spender
14        1       6.367797  Low_spender
15        1       6.367797  Low_spender
16        1       6.367797  Low_spender
17        1       6.367797  Low_spender
18        1       6.367797  Low_spender
19        1       6.367797  Low_spender

### Step 7

In [56]:
# Create a new column 'Customer_frequency'
ords_prods_merge['Customer_frequency'] = ords_prods_merge.groupby(['user_id'])['days_since_prior_order'].transform("median")

In [58]:
# Creating flag
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] > 20, 'frequency_flag'] = 'Non_frequent_customer'
ords_prods_merge.loc[(ords_prods_merge['Customer_frequency']>10) & (ords_prods_merge['Customer_frequency']<=20),'frequency_flag'] = 'Regular customer'
ords_prods_merge.loc[ords_prods_merge['Customer_frequency'] <= 10, 'frequency_flag'] = 'Frequent Customer'

In [64]:
# Check the new column
ords_prods_merge['frequency_flag'].value_counts(dropna = False)


frequency_flag
Frequent Customer        21559853
Regular customer          7208564
Non_frequent_customer     3636437
NaN                             5
Name: count, dtype: int64

NaN values indicate that there are 5 customers who have only made one purchase

In [67]:
ords_prods_merge[['user_id','days_since_prior_order', 'Customer_frequency', 'frequency_flag']].head(20)

user_id  days_since_prior_order  Customer_frequency         frequency_flag
0         1                     NaN                20.5  Non_frequent_customer
1         1                     NaN                20.5  Non_frequent_customer
2         1                     NaN                20.5  Non_frequent_customer
3         1                     NaN                20.5  Non_frequent_customer
4         1                     NaN                20.5  Non_frequent_customer
5         1                    15.0                20.5  Non_frequent_customer
6         1                    15.0                20.5  Non_frequent_customer
7         1                    15.0                20.5  Non_frequent_customer
8         1                    15.0                20.5  Non_frequent_customer
9         1                    15.0                20.5  Non_frequent_customer
10        1                    15.0                20.5  Non_frequent_customer
11        1                    21.0                20.5  Non_frequent_customer
12        1                    21.0                20.5  Non_frequent_customer
13        1                    21.0                20.5  Non_frequent_customer
14        1                    21.0                20.5  Non_frequent_customer
15        1                    21.0                20.5  Non_frequent_customer
16        1                    29.0                20.5  Non_frequent_customer
17        1                    29.0                20.5  Non_frequent_customer
18        1                    29.0                20.5  Non_frequent_customer
19        1                    29.0                20.5  Non_frequent_customer

### Exporting Data

In [ ]:
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'aggregated_ords_prods_merge.pkl'))